report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,97 @@
-evolution and refactoring
+### Refactoring
+
+The first task in this project was to refactor the provided system twice; first to Python 3, then a tech stack of the team's own choice.
+
+#### Python 2 to Python 3
+
+The conversion from Python 2 to 3 was almost symbolic for easy wins in DevOps.
+To do the conversion, a tool called [2to3](http://python3porting.com/2to3.html) was used; feed Python 2 in, get Python 3 out.
+
+It worked almost completely right away, which was surprising as the introduction to the tool in class said to be careful that nothing broke.
+Only thing needing a fix was the tests, which were previously dependent on the more raw string handling of Python 2, but that was not terribly hard to fix (See [This PR](https://github.com/jlndk/devoops/pull/6) for more).
+
+#### Python 3 with SQLite to ASP.NET with PostgreSQL
+
+The following conversion was much larger than the previous.
+The team decided to refactor the web server to an ASP.NET Core 3 application with PostgreSQL as the DBMS, as the language was familiar to some members, while being something stable to learn for the rest of the team.
+
+This refactor turned out to be quite monolithic, in part due to everything being changed at once.
+
+First part was difficult in and of itself, as ASP.NET provides at least 6 different ways to make a server-side rendered web page.
+The team settled on using a template built with Model-View-Controller architecture in mind, along with using ASP.NET Razor template pages for providing the HTML templates.
+
+Next part was to convert the Flask template pages to a format that Razor understands, which was simple enough, though tedious.
+Learning the syntax of Razor and Flask at the same time provided somewhat of a challenge, and porting to a MVC architecture at the same time didn't make this transition easier.
+
+The arguably biggest part was the implementation of Entity Framework (EF) and porting to PostgreSQL.
+For each part of the model, an entity would have to be established in code, and a context and repository layer for accessing and changing that entity would have to be written.
+Then, on top of that, an API layer was necessary to access and mutate this state.
+On the bright side, this completely eliminated plain text SQL queries in the code itself.
+
+If such a large refactor was to be done some other time, then a more incremental approach might be better.
+If the database was first changed to rely on another DBMS, and _then_ changed to work with EF, then the refactor could become more incremental and thus less monolithic.
+The only problem here is that it would equate to doing double the work, and when a refactor has to be done in a small timeframe, then double work doesn't seem practical.
```

Comment:
Maybe add that because we did all this work we didnt have any the next work. And it also made every change later easier.